### PR TITLE
New doc

### PR DESCRIPTION
**Pull request by CharlesAuthier**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,510 +1,135 @@
...
 
+## **Installation**
+Those step are for your a workstation on Ubuntu 18.04 using miniconda.
+Set and activate your python environment with the following commands:  
+```shell
+conda create -n gpu_ENV python=3.6 -c pytorch pytorch torchvision
+conda activate gpu_ENV
+conda install -c conda-forge ruamel_yaml h5py fiona rasterio geopandas scikit-image scikit-learn tqdm
+conda install -c fastai nvidia-ml-py3
+conda install mlflow
```
Comment (bstdenis) [inline README.md:48]: Missing segmentation_models_pytorch as dependency here. I installed it with
`pip install segmentation-models-pytorch`
but I see now that it's also on conda so that may be more consistent with the above...

File: README.md [hunk 2]
```diff
@@ -1,510 +1,135 @@
...
-To launch the program:  
+> The system can be used on your workstation or cluster and on [AWS](https://aws.amazon.com/).
 
+## **Installation**
+Those step are for your a workstation on Ubuntu 18.04 using miniconda.
+Set and activate your python environment with the following commands:  
+```shell
+conda create -n gpu_ENV python=3.6 -c pytorch pytorch torchvision
+conda activate gpu_ENV
+conda install -c conda-forge ruamel_yaml h5py fiona rasterio geopandas scikit-image scikit-learn tqdm
```
Comment (bstdenis) [inline README.md:46]: This line got stuck in solving environment for me (similar to https://github.com/conda/conda/issues/7690). I had to
`conda install --override-channels -c main -c conda-forge ruamel_yaml h5py fiona rasterio geopandas scikit-image scikit-learn tqdm`
to make it work. Perhaps this could be useful to other people in the future?

File: conf/config_template.yaml [hunk 3]
```diff
@@ -116,9 +115,7 @@ inference:
   chunk_size: 512 # (int) Size (height and width) of each prediction patch. Default: 512
   overlap: 10 # (int) Percentage of overlap between 2 chunks. Default: 10
   heatmaps: True
-
-
-# Visualization parameters
+  # Visualization parameters
   smooth_prediction: True
   overlap: 2 # overlap between tiles for smoothing. Must be an even number that divides chunk_size without remainder.
```
Comment (bstdenis) [inline conf/config_template.yaml:120]: overlap is used twice as a parameter in the inference section. I'm getting yaml warnings from this. Probably a bug.

File: utils/README.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+#### Intersection over Union (IoU)
+
+The IOU metric is a pixel based metric which measures overlap using the number of pixels common between the ground truth and the predictions divided by the total pixels across both.
+<p align="center">
+<img src="https://latex.codecogs.com/svg.latex?\Large&space;\textrm{IoU}=\frac{\textrm{ground truth} \cap \textrm{prediction}}{\textrm{ground truth} \cup \textrm{prediction}}" title="\Large IoU" class="center" />
+</p>
+
+#### Dice Similarity Coefficient  (dice)
+
+The dice metric scores model performance by measuring overlap between the ground truth and predictions divided by sum of pixels of both ground truth and predictions.
```
Comment (ymoisan) [inline utils/README.md:18]: I see `ground truth` and `ground-truth`.  1) Please be consistent and use only one.  2) I suggest "ground truth"

File: utils/README.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+<!-- classes = A and B evaluating for class A -->
+Where,
+- True Positives (TP) = pixels correctly classified as class A
+- False Positives (FP) = pixels incorrectly classified as class A
+- False Negatives (FN) = class A pixels incorrectly classified as class B
+- True Negatives (TN) = pixels correctly classified as class B
+
+#### Matthews Correlation Coefficient (MCC)
+
+The MCC metric takes all four confusion matrix categories (TP, FP, FN and TN) into account which in turn provides a more reliable score. This [article](https://bmcgenomics.biomedcentral.com/articles/10.1186/s12864-019-6413-7) compares MCC to other widely used metrics. This metric is insusceptible to the imbalanced dataset factor. Also [Scikit-Learn](https://scikit-learn.org/stable/modules/generated/sklearn.metrics.matthews_corrcoef.html) defines this metric is simpler terms.
```
Comment (ymoisan) [inline utils/README.md:45]: I know the word `insusceptible`exists, but I would rather say `not sensitive to` or even `not susceptible to` .  Could the *imbalanced dataset factor* be replaced by *class imbalance* ?

File: utils/README.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+</p>
+
+#### Accuracy
+Accuracy is simply defined here as the ratio of correctly classified pixels to total number of pixels between ground-truth and predicted.
+
+<p align="center">
+<img src="https://latex.codecogs.com/svg.latex?\Large&space;\textrm{acc}=\frac{\textrm{TP}+\textrm{TN}}{\textrm{TP}+\textrm{TN}+\textrm{FP}+\textrm{FN}}" title="\Large Acc" class="center" />
+</p>
+
+>**Note :** This metric is very susceptible to imbalanced datasets and may give an overly optimistic score when that is not the case. We can write down the accuracy formular using the confusion matrix categories.
```
Comment (ymoisan) [inline utils/README.md:57]: ... accuracy formula ...

File: utils/README.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+
+
+### Shape based Metrics
+
+**New shape based metrics would be added soon**
+
+---
+
+## **[Optimizers](optimizer.py)**
+For training a neural network to minimize the losses so as to perform better, we need to tweak the weights and parameters associated with the model and the loss function. This is where optimizers play a crucial role. Here some optimizers available in **GDL**.
```
Comment (ymoisan) [inline utils/README.md:75]: Last sentence : The following optimizers are ..."

File: conf/config_template.yaml (hunk 3 above)
Comment (CharlesAuthier) [reply conf/config_template.yaml:120]: Yeah its a new issues that we need to talk.

File: images_to_samples.py [hunk 8]
```diff
@@ -283,8 +283,42 @@ def samples_preparation(in_img_array,
...
 
+    Process
+    -------
+    1. Read csv file and validate existence of all input files and GeoPackages.
+
+    2. Do the following verifications:
+        1. Assert number of bands found in raster is equal to desired number
+           of bands.
+        2. Check that `num_classes` is equal to number of classes detected in
+           the specified attribute for each GeoPackage.
```
[the remark this answers is not in the document]
Comment (CharlesAuthier) [reply images_to_samples.py:295]: ??? this is not a doc on how to write a geopackage.

File: images_to_samples.py [hunk 9]
```diff
@@ -283,8 +283,42 @@ def samples_preparation(in_img_array,
...
+            - If gpkg's extent is bigger than raster's extent,
+              gpkg is clipped to raster's extent.
+        2. Convert GeoPackage vector information into the "label" raster with
+           utils.utils.vector_to_raster(). The pixel value is determined by the
+           attribute in the csv file.
+        3. Create a new raster called "label" with the same properties as the
+           input image.
+        4. Read metadata and add to input as new bands (*more details to come*).
+        5. Crop arrays in smaller samples of size `samples_size` and distance
+           `num_classes` specified in the configuration file. Visual
```
[the remark this answers is not in the document]
Comment (CharlesAuthier) [reply images_to_samples.py:313]: Sorry I didn't wrote this part I only copy past, but I will change it, thanks.

File: utils/README.md (hunk 4 above)
Comment (CharlesAuthier) [reply utils/README.md:18]: good thanks

**Review by remtav — APPROVED** [review]

**Review by victorlazio109 — COMMENTED** [review]
Great Job!

**Review by victorlazio109 — APPROVED** [review]

**Review by mpelchat04 — APPROVED** [review]

**Review by bstdenis — APPROVED** [review]

**Review by victorlazio109 — APPROVED** [review]
